torch_geometric/utils/augmentation.py

```diff
@@ -0,0 +1,204 @@
+        >>> x
+        tensor([[6., 7., 8.],
+                [3., 4., 5.],
+                [0., 1., 2.]])
+        >>> node_perm
+        tensor([2, 1, 0])
+    """
+    if not training:
+        perm = torch.arange(x.size(0), device=x.device)
+        return x, perm
+    perm = torch.randperm(x.size(0), device=x.device)
+    return x[perm], perm
+
+
+def mask_feature(x: Tensor, p: float = 0.5, mode: str = 'col',
+                 training: bool = True) -> Tuple[Tensor, Tensor]:
+    r"""Randomly masks feature from the feature matrix
+    :obj:`x` with probability :obj:`p` using samples from
+    a Bernoulli distribution.
+
+    The method returns (1) the retained :obj:`x`, (2) the feature
+    mask broadcastable with :obj:`x` (mode='row' and mode='col') or
+    with the same shape as :obj:`x` (mode='all'), indicating where features
+    are retained.
+
+    Args:
+        x (FloatTensor): The feature matrix.
+        p (float, optional): The masking ratio. (default: :obj:`0.5`)
+        mode (str, optional): The masked scheme to use for feature masking.
+            (:obj:`"row"`, :obj:`"col"` or :obj:`"all"`). (default: :obj:`col`)
+        training (bool, optional): If set to :obj:`False`, this operation is a
+            no-op. (default: :obj:`True`)
+
+    :rtype: (:class:`FloatTensor`, :class:`BoolTensor`)
+
+    Examples:
+
+        >>> # Masked features are column-wise sampled
+        >>> x = torch.tensor([[1, 2, 3],
+        ...                   [4, 5, 6],
+        ...                   [7, 8, 9]], dtype=torch.float)
+        >>> x, feat_mask = mask_feature(x)
+        >>> x
+        tensor([[1., 0., 3.],
+                [4., 0., 6.],
+                [7., 0., 9.]]),
+        >>> feat_mask
+        tensor([[True, False, True]])
+
+        >>> # Masked features are row-wise sampled
+        >>> x, feat_mask = mask_feature(x, mode='row')
+        >>> x
+        tensor([[1., 2., 3.],
+                [0., 0., 0.],
+                [7., 8., 9.]]),
+        >>> feat_mask
+        tensor([[True], [False], [True]])
+
+        >>> # Masked features are uniformly sampled
+        >>> x, feat_mask = mask_feature(x, mode='all')
+        >>> x
+        tensor([[0., 0., 0.],
+                [4., 0., 6.],
+                [0., 0., 9.]])
+        >>> feat_mask
+        tensor([[False, False, False],
+                [True, False,  True],
+                [False, False,  True]])
+    """
+    if p < 0. or p > 1.:
+        raise ValueError(f'Masking ratio has to be between 0 and 1 '
+                         f'(got {p}')
+    if not training or p == 0.0:
+        return x, torch.ones_like(x, dtype=torch.bool)
+    assert mode in ['row', 'col', 'all']
+
+    if mode == 'row':
+        mask = torch.rand(x.size(0), device=x.device) >= p
+        mask = mask.view(-1, 1)
+    elif mode == 'col':
+        mask = torch.rand(x.size(1), device=x.device) >= p
+        mask = mask.view(1, -1)
+    else:
+        mask = torch.randn_like(x) >= p
```

Comment:
Using `Tensor.bernoulli` to generate the mask looks more straightforward to me.

Reply:
But it is less efficient than simply using `torch.rand`, as suggested by https://github.com/pyg-team/pytorch_geometric/pull/5481#discussion_r976128424

Reply:
Something like
```suggestion
        mask = x.bernoulli(1-p).to(bool)
```
should be efficient?

Reply:
Not really sure about this. @rusty1s could you give a suggestion for us?

Reply:
When we want to use `torch.bernoulli`, we will create three intermediate tensors:
*  The input (`torch.full`)
* The output floating point tensor
* The mask
 
`torch.randn` only generates two:
* The random vector
* The mask

This optimization might not be necessary, but personally I feel they both read equally well.
Happy to go with `torch.bernoulli` if this improved readability if you strongly prefer it.

Reply:
Not sure if I missed something. I believe we do not need `torch.full` for generating masks.

```
from time import perf_counter
p = 0.2
x = torch.rand(1000, 1000)

t_start = perf_counter()
for _ in range(1000):
  mask = x.bernoulli(1-p).to(torch.bool)
t_stop = perf_counter()
print("Method 1 elapsed time:", t_stop-t_start)

t_start = perf_counter()
for _ in range(1000):
  mask = torch.randn_like(x) >= p
t_stop = perf_counter()
print("Method 2 elapsed time:", t_stop-t_start)
```
```

Method 1 elapsed time: 2.7997943980000173
Method 2 elapsed time: 7.306967894999957
```

It seems using `Tensor.bernoulli` is more efficient.

Reply:
Thank you @lightaime. Let's update the code and use `Tensor.bernoulli` instead.

Reply:
This is not entirely true. You need to move the creation of `x` to the for-loop.

Reply:
But seems to be still faster (which is weird). Happy to change it then :)